test/kubetest/oidc.go

```diff
@@ -0,0 +1,244 @@
+		}
+		return err
+	}
+}
+
+// CreateOIDCToken creates the token and associated materials required to execute a test
+func CreateOIDCToken(client kubernetes.Interface, token *OIDCTokenConfig, mockServerURL string) Action {
+	return func(ctx *ScenarioContext) error {
+		// Set up the token materials
+		jwk, jwt, err := createOIDCToken(token.Name, token)
+		if err != nil {
+			return fmt.Errorf("failed to create OIDC token: %w", err)
+		}
+
+		// Store the token in a secret
+		secretName := fmt.Sprintf("%s-token", token.Name)
+		_, err = client.CoreV1().Secrets(ctx.Namespace).Create(context.TODO(), &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: secretName,
+			},
+			Data: map[string][]byte{
+				"token.jwt": jwt,
+			},
+		}, metav1.CreateOptions{})
+		ctx.AddCleanUp(func() error {
+			return client.CoreV1().Secrets(ctx.Namespace).Delete(context.TODO(), secretName, metav1.DeleteOptions{})
+		})
+
+		var jwkEntry interface{}
+		err = json.Unmarshal(jwk, &jwkEntry)
+		if err != nil {
+			return fmt.Errorf("failed to unmarshal jwk into object: %w", err)
+		}
+
+		// Register an expectation with the mock-server for the /certs endpoint to return the JWK entry
+		mockClient := mockserver.NewClientURL(mockServerURL)
+		err = mockClient.RegisterExpectation(mockserver.Expectation{
+			Request: mockserver.RequestMatcher{
+				Method: http.MethodGet,
+				Path:   DefaultCertsURL,
+				Body: mockserver.BodyMatcher{
+					Type:      mockserver.MatchBodyJSON,
+					JSON:      nil,
+					MatchType: mockserver.TolerantMatch,
+				},
+			},
+			Response: mockserver.Response{
+				Body: map[string]interface{}{
+					"keys": []interface{}{jwkEntry},
+				},
+				StatusCode: 200,
+			},
+			Priority: 0,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to register /certs expectations with mock-issuer: %w", err)
+		}
+
+		return err
+	}
+}
+
+// createOIDCToken is a private utility that is capable of create a JWT and its associated JWK entry
+func createOIDCToken(name string, token *OIDCTokenConfig) ([]byte, []byte, error) {
+	// Set up the correct token type
+	signerOptions := jose.SignerOptions{}
+	signerOptions.WithType("JWT")
+
+	// Create a self-signed CA cert that can sign the token
+	cert, key, err := createSelfSignedCA(fmt.Sprintf("%s-token-signer", name))
+	signer, err := jose.NewSigner(jose.SigningKey{
+		Algorithm: jose.SignatureAlgorithm(token.Algorithm),
+		Key:       jose.JSONWebKey{Key: key, KeyID: token.KeyID},
+	}, &signerOptions)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create jwt signer: %s", err)
+	}
+
+	builder := jwt.Signed(signer)
+
+	claims := jwt.Claims{
+		Issuer:   token.Issuer,
+		Subject:  token.Subject,
+		IssuedAt: jwt.NewNumericDate(token.IssuedAt),
+		Expiry:   jwt.NewNumericDate(token.ExpiresAt),
+		Audience: token.Audience,
+	}
+
+	customClaims := CustomClaims{
+		Claims:   &claims,
+		Username: token.Username,
+		Roles:    token.Roles,
```

Comment:
I've got the group-claims mapped to roles on the command line arguments.   ...I've done this for no reason other than because this is how we use it in our project.